### PR TITLE
Add mechanism to set a width on withViewportMatch

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR adds a mechanism to make withViewportMatch usage evaluate as if the viewport had a given width.
This will allow us to dynamically change the with of the editor to simulate mobile usages and allows us now to make the editor appear in the customizer as it appears on mobile.

We create a new component ViewportMatchWidthProvider that accepts width as a property, and when used all with/ifViewportMatch descendants of it evaluate as if the viewPort had the specified width.

## How has this been tested?
I tested on the branch from this PR https://github.com/WordPress/gutenberg/pull/17960 (which uses this commit) and verified that I get a mobile-like UI in the customizer.

### CODE REVIEW

**Review by mapk — DISMISSED** [review]
Hey @jorgefilipecosta, thanks for continuing work on this. I'm not seeing the mobile view originally desired in the mockups. Am I missing something?

(_the colors below are off, so please ignore_)
Oct-30-2019 08-37-45

It should look more like this interface, https://github.com/WordPress/gutenberg/issues/16606, unless something has changed I'm not aware of.

**Comment by jorgefilipecosta** [comment]
Hi @mapk, two PR's are necessary to test this. We need to join the branch of this PR with the branch of PR https://github.com/WordPress/gutenberg/pull/17960.
To make things easier to test, I cherry-picked this commit into the branch of PR https://github.com/WordPress/gutenberg/pull/17960. So I think the best way to test this logic is to test PR https://github.com/WordPress/gutenberg/pull/17960 that also uses this logic.

File: packages/viewport/src/with-viewport-match.js [hunk 1]
```diff
@@ -33,11 +103,33 @@ import { withSelect } from '@wordpress/data';
...
+		return ( props ) => {
+			const globalQueriesResult = useSelect( ( select ) => {
+				return mapValues( queries, ( query ) => {
+					return select( 'core/viewport' ).isViewportMatch( query );
+				} );
+			}, [ queries ] );
+			const localQueries = useContext( ViewportMatchWidthContext );
+			const localQueriesResult = useMemo(
+				() => {
+					if ( ! localQueries ) {
```
Comment (aduth) [inline packages/viewport/src/with-viewport-match.js]: Some comments here to explain the logic would be useful to help clarify.

```suggestion
					// Only evaluate local queries if within a width context provider.
					if ( ! localQueries ) {
```

File: packages/viewport/src/with-viewport-match.js [hunk 2]
```diff
@@ -1,13 +1,83 @@
...
+};
+
+const ViewportMatchWidthContext = createContext( null );
+
+/**
+ * Component that makes all withViewportMatch usages descendents of its children,
+ * evaluate as if the viewport had the width specific as a prop.
+ */
+export const ViewportMatchWidthProvider = ( { width, children } ) => {
+	const queriesResult = useMemo(
```
Comment (aduth) [inline packages/viewport/src/with-viewport-match.js]: What is a "queries result" (shape, purpose)? I'm sure I could follow it along, but my initial impression is it's unclear what we're computing. An inline comment or JSDoc could help clarify.

Comment (aduth) [reply packages/viewport/src/with-viewport-match.js]: I almost wonder if it's even worth computing this cached value ahead of time. There are so few `withViewportMatch` (likely often zero within a provider?) and the operation seems reasonably trivial to compute on the fly within `withViewportMatch` than to compute all possible options ahead of time. Also keeps things a little easier to follow for that logic to be colocated.

File: packages/viewport/src/with-viewport-match.js [hunk 3]
```diff
@@ -1,13 +1,83 @@
...
+		() => {
+			return reduce(
+				BREAKPOINTS,
+				( resultBreakpointsReduce, breakpointValue, breakpointName ) => {
+					return reduce(
+						OPERATORS,
+						( resultOperatorsReduce, operatorMediaQuery, operator ) => {
+							const operatorKey = `${ operator } ${ breakpointName }`;
+							const evaluationResult = OPERATOR_EVALUATORS[ operator ](
+								breakpointValue,
```
Comment (aduth) [inline packages/viewport/src/with-viewport-match.js]: This nesting gets pretty excessive and becomes hard to follow.

I might suggest:

- Not every argument needs to be on its own line
- Consolidate the operations.

That being said, since we're combining an aggregate of two different objects, I guess it's inevitable that we'll have some nesting like this.

File: packages/viewport/src/with-viewport-match.js [hunk 4]
```diff
@@ -33,11 +103,33 @@ import { withSelect } from '@wordpress/data';
...
+					return mapValues( queries, ( query ) => {
+						return localQueries[ query ];
+					} );
+				},
+				[ localQueries, queries ]
+			);
+			return (
+				<WrappedComponent
+					{ ...props }
+					{ ...( localQueriesResult ? localQueriesResult : globalQueriesResult ) }
```
Comment (aduth) [inline packages/viewport/src/with-viewport-match.js]: I'm not sure if there are options, but it seems wasteful to compute _both_ the local and global results, only to ultimately pass just one along.

The only thing I could consider is if we extracted separate subcomponents for each whether there is or isn't context, but I suspect that extra layer might just make things more convoluted anyways.

```jsx
function GlobalViewportResult( { props } ) {}

function ContextViewportResult( { props } ) {}

return ( props ) => {
	const width = useContext( ViewportMatchWidthContext );
	return width ? 
		<ContextViewportResult { ...props } width={ width } /> : 
		<GlobalViewportResult { ...props } />;
};
```

🤷‍♂

**Comment by jorgefilipecosta** [comment]
Hi @aduth thank you for the review. This PR was updated applying the suggestions you provided and a unit test was added so we can easily test the expected behavior.

File: packages/viewport/src/with-viewport-match.js [hunk 5]
```diff
@@ -6,8 +6,69 @@ import { mapValues } from 'lodash';
...
+	'>=': ( breakpointValue, width ) => ( width >= breakpointValue ),
+};
+
+const ViewportMatchWidthContext = createContext( null );
+
+/**
+ * Component that makes all withViewportMatch usages descendents of its children,
+ * evaluate as if the viewport had the width specific as a prop.
+ */
+export const ViewportMatchWidthProvider = ViewportMatchWidthContext.Provider;
```
Comment (youknowriad) [inline packages/viewport/src/with-viewport-match.js]: Now that we have `useViewportMatch`, do you think we should move that provider to the compose package instead?

Comment (jorgefilipecosta) [reply packages/viewport/src/with-viewport-match.js]: Yes, I'm working on this. The uncertainty I'm finding is if we should also expose the consumer of the context to allow it to be used in simulations by withViewPortMatch. I wanted a single provider to simulate both in the hook and HOC.

Comment (youknowriad) [reply packages/viewport/src/with-viewport-match.js]: Another possibility is to refactor our existing class components that use the HoC to functional components and use the hook everywhere.

Comment (jorgefilipecosta) [reply packages/viewport/src/with-viewport-match.js]: It seems we still have lots of usages of withViewPortMatch and indirect usages via ifViewportMatch. I guess we can try to refactor all of those, but it will take some time. Another con of this solution is that some blocks may be using withViewPortMatch, and in that case, these use cases will not be simulated.
I guess we can simply expose the simulation via a hook: useViewportMatch.__experimentalUseSimulatedValue()

Then we can use the hook inside the withViewPortMatchSimulation to retrieve the simulated value from the context.

@youknowriad what you think of this idea?

Comment (youknowriad) [reply packages/viewport/src/with-viewport-match.js]: Not sure I follow though.

One viable approach too is to rewrite withViewportMatch to use useViewportMatch under the hood. that way we only have to support the simulation in the hook.

Comment (jorgefilipecosta) [reply packages/viewport/src/with-viewport-match.js]: Basically I was proposing an experimental hook that exposes the width simulation value from the context
```
const simulationWidth = seViewportMatch.__experimentalUseSimulatedValue();

if( simulationWidth ) {...
```

But I think making withViewPortMatch use the hook is a better idea and it is something we should do anyway because we need to keep withViewPortMatch for back-compatibility and doing this we converge to a single codebase (way of doing things).

**Comment by jorgefilipecosta** [comment]
Hi @aduth, @youknowriad,
This PR was updated to make the simulation happen on useViewportMatch (happening by consequence on withViewPort match because the HOC uses the hook).
The changes in https://github.com/WordPress/gutenberg/blob/add/mechanism-to-set-a-width-on-withViewportMatch/packages/edit-post/src/editor.js are for testing purposes and will not be merged.
To test I changed values on this file and verified the flag passed to the image block was the expected one.

**Comment by jorgefilipecosta** [comment]
This PR was updated and unit tests were added, should be ready for a review.

File: packages/compose/src/hooks/use-viewport-match/index.js [hunk 6]
```diff
@@ -37,6 +42,13 @@ const CONDITIONS = {
 	'<': 'max-width',
 };
 
+const OPERATOR_EVALUATORS = {
+	'<': ( breakpointValue, width ) => ( width < breakpointValue ),
+	'>=': ( breakpointValue, width ) => ( width >= breakpointValue ),
+};
+
+const ViewportMatchWidthContext = createContext( null );
```
Comment (youknowriad) [inline packages/compose/src/hooks/use-viewport-match/index.js]: Maybe just ViewportWidth context.

**Review by youknowriad — COMMENTED** [review]
This looks good but I was wondering, this only affects the viewportMatch behavior (JS checks) and I believe we have another mechanism to do the same for CSS right?

could we combine those somehow? Meaning if we add the provider to a React tree, the provider also triggers the CSS forcing inside that tree?

**Comment by jorgefilipecosta** [comment]
> This looks good but I was wondering, this only affects the viewportMatch behavior (JS checks) and I believe we have another mechanism to do the same for CSS right?

> could we combine those somehow? Meaning if we add the provider to a React tree, the provider also triggers the CSS forcing inside that tree?


Hi @youknowriad, I guess the mechanisms make sense used together but I separated in two PR's to make reviewing easier. My plan is to provide a single component that uses the mechanism to simulate media queries and the mechanism to simulate the value on the hook (uses both components to combine the mechanisms).

**Review by youknowriad — APPROVED** [review]

File: packages/compose/src/hooks/use-media-query/index.js [hunk 7]
```diff
@@ -6,12 +6,15 @@ import { useState, useEffect } from '@wordpress/element';
 /**
  * Runs a media query and returns its value when it changes.
  *
- * @param {string} query Media Query.
+ * @param {?string} query Media Query.
```
Comment (aduth) [inline packages/compose/src/hooks/use-media-query/index.js]: Nit: If it's optional, this should be:

```suggestion
 * @param {string} [query] Media Query.
```

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#nullable-undefined-and-void-types

File: packages/compose/src/hooks/use-viewport-match/index.js [hunk 8]
```diff
@@ -53,8 +65,15 @@ const CONDITIONS = {
...
+	const simulatedWidth = useContext( ViewportMatchWidthContext );
+	const mediaQuery = ! simulatedWidth && `(${ CONDITIONS[ operator ] }: ${ BREAKPOINTS[ breakpoint ] }px)`;
+	const mediaQueryResult = useMediaQuery( mediaQuery );
+	if ( simulatedWidth ) {
+		return OPERATOR_EVALUATORS[ operator ]( BREAKPOINTS[ breakpoint ], simulatedWidth );
+	}
+	return mediaQueryResult;
 };
 
+useViewportMatch.__experimentalWidthProvider = ViewportMatchWidthContext.Provider;
```
Comment (aduth) [inline packages/compose/src/hooks/use-viewport-match/index.js]: Is there any prior art for how we assign context as properties of hooks? Wondering if it should be its own exported member instead.

File: packages/compose/src/hooks/use-media-query/index.js [hunk 9]
```diff
@@ -6,12 +6,15 @@ import { useState, useEffect } from '@wordpress/element';
...
  * @return {boolean} return value of the media query.
  */
 export default function useMediaQuery( query ) {
 	const [ match, setMatch ] = useState( false );
 	useEffect( () => {
+		if ( ! query ) {
```
Comment (aduth) [inline packages/compose/src/hooks/use-media-query/index.js]: I get that hooks need to be rendered deterministically, but it still feels odd to me that we'd want to support this use case of an empty `query` for `useMediaQuery`. Do the default hooks support this practice? i.e. `useCallback( undefined )` etc? Are there patterns in the hooks landscape for this sort of "forking" logic (i.e. do different behavior based on some condition)?

File: packages/compose/src/hooks/use-viewport-match/index.js [hunk 10]
```diff
@@ -37,6 +42,13 @@ const CONDITIONS = {
 	'<': 'max-width',
 };
 
+const OPERATOR_EVALUATORS = {
```
Comment (aduth) [inline packages/compose/src/hooks/use-viewport-match/index.js]: Can we document these?

Comment (jorgefilipecosta) [reply packages/compose/src/hooks/use-viewport-match/index.js]: The object was documented.

File: packages/compose/src/hooks/use-media-query/index.js (hunk 9 above)
Comment (jorgefilipecosta) [reply packages/compose/src/hooks/use-media-query/index.js]: Hi @aduth, I verified that useCallback( undefined ) returns undefined useCallback( false ) returns false, this behavior is equivalent to what we did useMediaQuery.
Without this pattern, I am not aware of any way we could implement this hook and avoid calling the media query listeners even when they are not needed because the value is simulated.

File: packages/compose/src/hooks/use-viewport-match/index.js (hunk 8 above)
Comment (jorgefilipecosta) [reply packages/compose/src/hooks/use-viewport-match/index.js]: No, I think we never did this before. I exported as a property of the hook to make clear that the component only affects the hook return value. But I can export an __experimentalUseViewportMatchWidthProvider  component instead. I don't have a preference here. If you have a preference for the component version, let me know, and I will perform the update.

Comment (aduth) [reply packages/compose/src/hooks/use-viewport-match/index.js]: Hm. The main thing that caught my eye was in how it's used:

```jsx
<useViewportMatch.__experimentalWidthProvider value={ undefined } />
```

Technically it's valid, but a little non-conventional.

For me, if it was a separate thing, I don't think we would want to associate it so strongly to being only relevant for this hook, so I would imagine something like `ViewportWidthProvider` being sufficient.

But especially given that this package is meant to be used exclusively for these sorts of helpers (higher-order components and hooks), I think it would be fair to associate it directly with the hook.

For that reason, I think what you've proposed here can be okay.

File: packages/edit-post/src/editor.js [hunk 11]
```diff
@@ -116,20 +116,22 @@ class Editor extends Component {
...
-								{ ...props }
-							>
-								<ErrorBoundary onError={ onError }>
-									<EditorInitialization postId={ postId } />
-									<Layout />
-									<KeyboardShortcuts shortcuts={ preventEventDiscovery } />
-								</ErrorBoundary>
-								<PostLockedModal />
-							</EditorProvider>
+							<useViewportMatch.__experimentalWidthProvider value={ undefined }>
```
Comment (aduth) [inline packages/edit-post/src/editor.js]: Do we need to force that this provider be rendered in the application? Isn't it enough for `useContext` to use a default when there's no explicit context value?

It would make the hook easier to use in other applications. And, unless I'm mistaken, it should "just work" if this wrapper element were removed here.

Comment (jorgefilipecosta) [reply packages/edit-post/src/editor.js]: Hi @aduth, this file is not supposed to be merged. I just added this here to make it easier to test, one can just add value here and see if the value is applied and the editor reacts differently.

Comment (aduth) [reply packages/edit-post/src/editor.js]: > Hi @aduth, this file is not supposed to be merged.

But it was merged? 🤔

Comment (aduth) [reply packages/edit-post/src/editor.js]: Oh, I see #19075 now. Will review 👍

Comment (jorgefilipecosta) [reply packages/edit-post/src/editor.js]: Yes, I'm sorry for that I prepared the revert and missed the push :(
